Data loss when using `dev-file` DB with H2 2.3.232

### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

storage

### Describe the bug

A regression was introduced with the [Quarkus 3.14 upgrade](https://github.com/keycloak/keycloak/pull/32519). Sometime, data are not persistent between server restarts when using file based H2. I observed whole Realms missing.

This also causes: https://github.com/keycloak/keycloak/issues/32784

### Version

main

### Regression

- [X] The issue is a regression

### Expected behavior

Records in the DB are correctly persisted and survive server restarts.

### Actual behavior

Records are missing after restart.

### How to Reproduce?

* [Quick and dirty reproducer test](https://github.com/vmuzikar/keycloak/blob/81280350a0a82666c20d7959dc914c8f0c450b88/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/H2Reproducer.java)
* [CI flow](https://github.com/vmuzikar/keycloak/blob/81280350a0a82666c20d7959dc914c8f0c450b88/.github/workflows/ci.yml)

### Anything else?

Few observations:
* Realms are definitely affected, not sure about other tables.
* When connecting to the DB externally when the server is shut down, it seemed the records are sometimes missing just after the shut down, other times after a next server start.
* It seems to happen when DB is updated just before the server shutdown.
* It is not a timing issue. Giving the server more time to write transactions into the persistent storage does not help.
* Missing are even records that survived previous server restart and were untouched since then. So it doesn't affect just last transactions that were not written to hard drive in time.
* Tracing H2 log doesn't seem to contain anything suspicious (log are part of the [CI flow](https://github.com/vmuzikar/keycloak/actions?query=branch%3Aflaky-hostname) in the reproducer, see for example [this artifact](https://github.com/vmuzikar/keycloak/actions/runs/10797461621/artifacts/1915660150)).
* Seems to affect only H2.
* Behaves the same regardless how the server is killed (`Process#destroy`, dedicated `kill` shell command).
* Explicitly killing process descendants did not help.
* Did not yet confirmed reproducibility outside of Keycloak.
* H2 options tried without any effect:
  * `WRITE_DELAY=0`
  * `DB_CLOSE_DELAY=-1`
  * `DB_CLOSE_ON_EXIT=FALSE`
  * `AUTO_SERVER=TRUE`
  * `FILE_LOCK=SOCKET`
* We can't use older H2 versions with newer Quarkus due to explicit validations for [baked-in H2 version](https://github.com/quarkusio/quarkus/blob/3.14.2/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/config/DialectVersions.java#L26).
* ~~Overall, seems like H2 issue.~~ It is not reproducible with Undertow, likely it is a Quarkus issue.

Reproduced the issue for the upgrade pr case with the detailed logging enabled - https://github.com/shawkins/keycloak/actions/runs/10818060035/job/30013072936 - the logs are massive, so we'll probably want to pare things down as we get further.

The error manifested as @vmuzikar described. Lot's of realms exist exist - for quite a while - then disappear. This snippet shows up to realm 22 was created at one point, then ~40 seconds later only up to realm 9 exists.

```
2024-09-11T19:31:12.3399624Z 2024-09-11 19:25:00,798 DEBUG [org.hibernate.internal.util.EntityPrinter] (executor-thread-1) org.keycloak.models.jpa.entities.ClientEntity{publicClient=false, registeredNodes=null, authFlowBindings=null, description=null, alwaysDisplayInConsole=false, serviceAccountsEnabled=false, secret=null, consentRequired=false, clientAuthenticatorType=client-secret, enabled=true, notBefore=0, frontchannelLogout=false, surrogateAuthRequired=false, protocol=null, bearerOnly=true, nodeReRegistrationTimeout=0, managementUrl=null, id=e760000b-b946-4a8c-95ad-394c7feca6e5, clientId=realm-22-realm, realmId=838b4686-8fa6-4955-a3f2-19076ed284a4, protocolMappers=[], redirectUris=null, fullScopeAllowed=false, rootUrl=null, directAccessGrantsEnabled=false, implicitFlowEnabled=false, registrationToken=null, baseUrl=null, name=realm-22 Realm, standardFlowEnabled=true, attributes=[], webOrigins=null, scopeMappingIds=null}
...
2024-09-11T19:31:53.9410626Z java.lang.AssertionError: Realms did not match expected:<[2b38598c-d261-4629-9043-f9c4f1d2d998:realm-4, 80bd2553-e0ee-4a17-83b4-a3141a75b06a:realm-22, 59afbccc-1c53-4cb8-9fb0-ea0213c1d8c1:realm-5, 91df25bd-2452-4dfb-a405-dde9fe98ab0d:realm-6, a4306a37-e2e0-4c12-9a1c-f9bbb9734df4:realm-12, bdd1dbc0-5580-4404-9c3e-59328adcd5fe:realm-2, 92a63d52-1f8a-45bb-b1e2-78ad894917fd:realm-11, f52b10d8-eb17-4605-b183-7c745e0339b1:realm-21, 9e3bc463-1f2f-4c01-859b-e085ccf8e33a:realm-9, d2043f60-b00e-4de4-9e03-18946c1c53e9:realm-15, fc2d70df-ea9f-4012-92af-3100fa380194:realm-7, 1d8ac448-93c8-4b02-a159-5928dd2ea4fe:realm-16, 321635b6-4ce6-4887-8bd7-1ce3bb362f68:realm-8, 252de93b-a826-46ad-9076-8235a36b605a:realm-14, 2daebb43-99a4-4d1c-9814-8403cfe89b3b:realm-18, 589fac9a-bc32-4564-a02d-b63a64d247b8:realm-20, 86adc6b9-3c81-403e-ae70-5b4b7cdef4ab:realm-13, a2015bf7-cef9-4130-a5ab-e3c7227cb6a0:realm-19, 04b47ca5-78c1-4dfc-bced-0babffe54ff1:realm-10, 227d361e-dead-4d9e-91a2-5c41c7ddbade:realm-1, 7d0043e7-050a-460d-be89-98b6b6040ef8:realm-0, ca183adf-ac54-405d-890f-b9d297b17dc1:realm-3, 838b4686-8fa6-4955-a3f2-19076ed284a4:master, 96b35244-4621-463c-a500-7c0be224916a:realm-17]> but was:<[321635b6-4ce6-4887-8bd7-1ce3bb362f68:realm-8, 2b38598c-d261-4629-9043-f9c4f1d2d998:realm-4, 59afbccc-1c53-4cb8-9fb0-ea0213c1d8c1:realm-5, 91df25bd-2452-4dfb-a405-dde9fe98ab0d:realm-6, bdd1dbc0-5580-4404-9c3e-59328adcd5fe:realm-2, 227d361e-dead-4d9e-91a2-5c41c7ddbade:realm-1, 7d0043e7-050a-460d-be89-98b6b6040ef8:realm-0, ca183adf-ac54-405d-890f-b9d297b17dc1:realm-3, 838b4686-8fa6-4955-a3f2-19076ed284a4:master, 9e3bc463-1f2f-4c01-859b-e085ccf8e33a:realm-9, fc2d70df-ea9f-4012-92af-3100fa380194:realm-7]>
```

--- a/pom.xml
+++ b/pom.xml
@@ -51,8 +51,8 @@
         <jboss.snapshots.repo.id>jboss-snapshots-repository</jboss.snapshots.repo.id>
         <jboss.snapshots.repo.url>https://s01.oss.sonatype.org/content/repositories/snapshots/</jboss.snapshots.repo.url>
 
-        <quarkus.version>3.14.2</quarkus.version>
-        <quarkus.build.version>3.14.2</quarkus.build.version>
+        <quarkus.version>3.14.4</quarkus.version>
+        <quarkus.build.version>3.14.4</quarkus.build.version>
 
         <project.build-time>${timestamp}</project.build-time>
 
@@ -92,7 +92,7 @@
         <bouncycastle.bctls-fips.version>1.0.19</bouncycastle.bctls-fips.version>
 
         <dom4j.version>2.1.3</dom4j.version>
-        <h2.version>2.3.232</h2.version>
+        <h2.version>2.3.230</h2.version>
         <hibernate-orm.plugin.version>6.2.13.Final</hibernate-orm.plugin.version>
         <hibernate.c3p0.version>6.2.13.Final</hibernate.c3p0.version>
         <infinispan.version>15.0.8.Final</infinispan.version>

